relint.el

```diff
@@ -1232,6 +1232,90 @@ or in the car of an element."
         (relint--check-re-string (car elem) ident pos p)))))
    form path))
 
+(defun relint--check-treesit-query-sexp (expr name pos path)
+  "Recursively check tree-sitter :match regexps in EXPR.
+EXPR is assumed a non-literal in the source."
+  (pcase expr
+    (`(,_ . ,(or `((:match ,(and (pred stringp) re) . ,_))
+                 ;; Optional capture name.
+                 `(,(pred symbolp)
+                   (:match ,(and (pred stringp) re) . ,_))))
+     (relint--check-re-string re name pos path))
+    ((pred consp)
+     (while (consp expr)
+       (relint--check-treesit-query-sexp (car expr) name pos path)
+       (setq expr (cdr expr))))
+    ((pred vectorp)
+     (dotimes (i (length expr))
+       (relint--check-treesit-query-sexp (aref expr i) name pos path)))))
+
+(defun relint--check-treesit-query (form name pos path)
+  "Check tree-sitter #match and :match regexps in FORM.
+FORM is assumed a non-literal in the source."
+  (if (not (stringp form))
+      (relint--check-treesit-query-sexp form name pos path)
+    (let ((i 0))
+      (while (string-match (rx "(" (* (any " \n\t")) "#match") form i)
+        (setq i (match-end 0))
+        (let ((re (ignore-errors (car (read-from-string form i)))))
+          (when (stringp re)
+            (relint--check-re-string re name pos path)))))))
```

Comment:
Not too sure about this part: the string we're inspecting does contain an S-expression, but it doesn't follow Emacs s-exp syntax exactly so there could be problems.

Normally such a cavalier attitude to rules and specs would be completely in character for relint, but in this case there does simply not seem to be much to gain: the Emacs tree does not contain a single use of string-form queries as far as I can tell (outside the test that checks that it works). Exposing the tree-sitter sexp syntax to the user at all seems to have been a mistake.

In fact I'm somewhat worried about `treesit-query-expand` and `treesit-pattern-expand` which happily use `prin1-to-string` to produce a tree-sitter s-exp. This seems completely wrong: try inserting a newline escape sequence, `\n`, in the regexp of an otherwise valid query (in Emacs s-exp form).

Which is good, because it means that merely by proposing this patch you have already uncovered bugs in Emacs. Next!

Reply:
For posterity, this is https://bugs.gnu.org/64017.

Shall I remove the string query syntax checking from `relint` in the meantime?

Reply:
Would you extend the test coverage a little, at least in a kind of representative way? I'm not aiming for full coverage but to be reasonably well defended against unintentional damage to the code.

Then if you rebase the patches, I'll give it a final look-over and merge.